znoyder/mapper.py

```diff
@@ -0,0 +1,135 @@
+#!/usr/bin/env python3
+#
+# Copyright 2022 Red Hat, Inc.
+# All Rights Reserved.
+#
+#    Licensed under the Apache License, Version 2.0 (the "License"); you may
+#    not use this file except in compliance with the License. You may obtain
+#    a copy of the License at
+#
+#         http://www.apache.org/licenses/LICENSE-2.0
+#
+#    Unless required by applicable law or agreed to in writing, software
+#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+#    License for the specific language governing permissions and limitations
+#    under the License.
+#
+
+from copy import deepcopy
+import re
+
+from znoyder.config import copy_map
+from znoyder.config import exclude_map
+from znoyder.config import include_map
+from znoyder.config import override_map
+from znoyder.lib.zuul import ZuulJob
+
+
+def match(string: str, specifier: str) -> bool:
+    '''Function checks if a given string is matched by a given specifier.
+
+    The match is performed in the awk-inspired fashion:
+    – if the specifier starts and ends with the forward slash (/), e.g. /foo/,
+      then the content between slashes is treated as regular expression,
+    – otherwise it is a value that should fully match the input string.
+
+    Parameters
+    ----------
+    string : str
+        The string that should be tested against specifier.
+    specifier : str
+        The expected value or regular expression to be matched against.
+
+    Returns
+    -------
+    matched : bool
+        True if given string matches the specifier, False otherwise.
+
+    Examples
+    --------
+    >>> match('foobar', 'foobar')
+    True
+    >>> match('foobar', 'foo')
+    False
+    >>> match('foobar', '/foo/')
+    True
+    '''
+
+    if specifier.startswith('/') and specifier.endswith('/'):
+        regex = re.compile(specifier[1:-2])
+        return bool(regex.search(string))
+    else:
+        regex = re.compile(specifier)
+        return bool(regex.fullmatch(string))
+
+
+def job_from_map_entry(entry: dict) -> ZuulJob:
+    job_name, job_options = deepcopy(entry)
+    job_trigger_type = job_options.pop('type', 'check')
+
+    if isinstance(job_trigger_type, str):
+        return [ZuulJob(job_name, job_trigger_type, job_options)]
+    else:
+        return [ZuulJob(job_name, trigger_type, job_options)
+                for trigger_type in job_trigger_type]
+
+
+def exclude_jobs(jobs, project, tag) -> list:
+    included_jobs = []
+
+    for job in jobs:
+        wanted = True
+
+        for project_specifier in exclude_map:
+            if not match(project, project_specifier):
+                continue
+
+            for tag_specifier in exclude_map[project_specifier]:
```

Comment:
I would maybe move the nested loop to a separate function, so that we can return and no keep track of the state with a variable and multiple breaks

Reply:
so, you propose to define here a local functions like `match_project` and `match_tag`?
I will see how it will look like.
At the same time, I am considering to reverse the logic here, because now we fill the variable with entries that we do not want to exclude (this is translated what it was orignally here with previous maps), which a little bit way around the problem.

Thanks for pointing that out, I will think of it for a while and propose something cleaner.

Reply:
Yes, something like `should_include_project(job, exclude_map[project_specifier], tag_specifier)` that would return a bool with whether it should be included.

Reply:
@cescgina I attempted to implement that, but maybe I am too stupid, however whatever I produced did not look better/easier to understand, so instead I just rewritten that with more natural logic (filter out element matched by exclude map). IMO It seems easier to read now and pretty straightforward without all those breaks around – can you recheck?